### PR TITLE
Add changes for AVX-512 support in k-NN. 

### LINKED ISSUE #2056: [FEATURE] Add support for FAISS AVX512
**Is your feature request related to a problem?**
Improve performance of vector search and indexing when using the FAISS library, by using AVX512 accelerators which are supported in faiss. 

**What solution would you like?**
Add support to use AVX512, initially as an optional feature and eventually enable by default on machines where AVX512 is supported on subsequent OpenSearch releases.
This will be a 2 part change: 
1) Adds necessary code in k-NN to check for AVX512 support on the machine.
2) A separate change made in FAISS to add missing functions support for AVX512.

### PR DESCRIPTION
**Pull request by akashsha1**
### Description
This change adds support to speed up vector search and indexing in faiss using AVX512 hardware accelerator.

### Related Issues
Resolves #2056 

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose). https://github.com/opensearch-project/documentation-website/pull/8307

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by assanedi** [comment]
Benchmark was run using opensearch-benchmark with cohere dataset(768 dimensions). 
The opensearch cluster was deployed with 2 data nodes (r7i.2xlarges),  1 replica and 4 shards. 
Using this setup AVX512 shows 15% improvement over AVX2 as shown below: 

image

**Comment by naveentatikonda** [comment]
> Benchmark was run using opensearch-benchmark with cohere dataset(768 dimensions). The opensearch cluster was deployed with 2 data nodes (r7i.2xlarges), 1 replica and 4 shards. Using this setup AVX512 shows 15% improvement over AVX2 on indexing and 7 % on search as shown below:
> 
> image

@assanedi Can you also pls add other configuration details like the indexing clients, query clients, ef_construction, ef_search, etc

**Review by naveentatikonda — DISMISSED** [review]
LGTM! Thanks @akashsha1 @assanedi

**Comment by naveentatikonda** [comment]
> "target_index_bulk_indexing_clients": 20,
"target_index_max_num_segments": 10,
"hnsw_ef_search": 256,
"hnsw_ef_construction": 256

@assanedi Isn't the max_num_segments was 1 during forcemerge ?

**Comment by assanedi** [comment]
> > "target_index_bulk_indexing_clients": 20,
> > "target_index_max_num_segments": 10,
> > "hnsw_ef_search": 256,
> > "hnsw_ef_construction": 256
> 
> @assanedi Isn't the max_num_segments was 1 during forcemerge ?

Yes I run the forcemerge API, here is the results of it:
curl -X POST -k --user admin:admin http://10.0.0.80:9200/_forcemerge?max_num_segments=1
{"_shards":{"total":8,"successful":8,"failed":0}}

**Comment by naveentatikonda** [comment]
> 
> Yes I run the forcemerge API, here is the results of it: curl -X POST -k --user admin:admin http://10.0.0.80:9200/_forcemerge?max_num_segments=1 {"_shards":{"total":8,"successful":8,"failed":0}}

Yes, but in the configuration you mentioned it as 10 instead of 1 for `target_index_max_num_segments `

File: DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -283,19 +283,22 @@ make -j 4
...
+by enabling `AVX2` and `AVX512` on `x86 architecture` and `NEON` on `ARM64 architecture` where applicable while building the Faiss library. But to enable SIMD,
+the underlying processor should support these capabilities (AVX512, AVX2 or NEON). It can be disabled by setting the parameter `avx2.enabled` to `false` and
+`avx512.enabled` to `false`. If your processor supports `AVX512` or `AVX2`, they can be set by enabling the setting . By default, these values are enabled on 
+OpenSearch. Some exceptions: As of now, SIMD support is not supported on Windows OS, and AVX512 is not present on MAC systems due to hardware not supporting the
+feature.
 
 ```
 # While building OpenSearch k-NN
-./gradlew build -Dsimd.enabled=true
+./gradlew build -Davx2.enabled=true -Davx512.enabled=true
```
Comment (jmazanec15) [inline DEVELOPER_GUIDE.md:294]: Shouldnt these be mutually exclusive?

**Comment by naveentatikonda** [comment]
For FP32 we don’t need to make any changes in Faiss as they are using auto-vectorization to achieve the optimization with AVX512. But, for Scalar Quantization Intel have raised a PR to Faiss which is under review
https://github.com/facebookresearch/faiss/pull/3853

File: src/main/java/org/opensearch/knn/index/KNNSettings.java [hunk 2]
```diff
@@ -499,6 +512,22 @@ public static boolean isFaissAVX2Disabled() {
         }
     }
 
+    public static boolean isFaissAVX512Disabled() {
+        try {
+            return KNNSettings.state().getSettingValue(KNNSettings.KNN_FAISS_AVX512_DISABLED);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/KNNSettings.java:517]: Can we do proper null checks here? In general, I think its best to avoid catching all exceptions.

File: src/main/java/org/opensearch/knn/jni/FaissService.java [hunk 3]
```diff
@@ -30,14 +32,17 @@
  *      src/main/java/org/opensearch/knn/index/query/KNNQueryResult.java
  *      src/main/java/org/opensearch/knn/common/KNNConstants.java
  */
+@SuppressWarnings("deprecation")
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/jni/FaissService.java:35]: Why was this needed?

File: src/main/java/org/opensearch/knn/jni/PlatformUtils.java [hunk 4]
```diff
@@ -80,4 +81,38 @@ public static boolean isAVX2SupportedBySystem() {
         }
         return false;
     }
+
+    public static boolean isAVX512SupportedBySystem() {
+
+        if (!Platform.isIntel() || Platform.isMac() || Platform.isWindows()) {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/jni/PlatformUtils.java:87]: does mac support avx2 but not avx512?

File: src/main/java/org/opensearch/knn/jni/PlatformUtils.java [hunk 5]
```diff
@@ -80,4 +81,38 @@ public static boolean isAVX2SupportedBySystem() {
...
+
+            try {
+                return AccessController.doPrivileged((PrivilegedExceptionAction<Boolean>) () -> {
+                    Stream<String> linestream = Files.lines(Paths.get(fileName));
+                    String flags = linestream.filter(line -> line.startsWith("flags")).findFirst().orElse("");
+                    return Arrays.stream(avx512).allMatch(flags::contains);
+                });
+
+            } catch (Exception e) {
+                logger.error("[KNN] Error reading file [{}]. [{}]", fileName, e.getMessage(), e);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/jni/PlatformUtils.java:113]: Same as above

File: src/main/java/org/opensearch/knn/jni/PlatformUtils.java (hunk 4 above)
Comment (akashsha1) [reply src/main/java/org/opensearch/knn/jni/PlatformUtils.java:87]: that's right. mac moved away from x86 right around the time avx512 was introduced. 
I don't have mac systems to verify this, and @naveentatikonda confirmed his mac system doesn't have avx512. 
if you know of any mac systems supporting avx512, I can update the check here.

File: src/main/java/org/opensearch/knn/jni/FaissService.java (hunk 3 above)
Comment (akashsha1) [reply src/main/java/org/opensearch/knn/jni/FaissService.java]: added by VSCode accidentally. Fixed with the latest commit.

File: src/main/java/org/opensearch/knn/jni/PlatformUtils.java (hunk 5 above)
Comment (akashsha1) [reply src/main/java/org/opensearch/knn/jni/PlatformUtils.java:114]: updated to PrivilegedActionException. 
However, do note that AccessController API itself is deprecated by Java, and hence the IDE prompts the warning. This'll require a broader change across the codebase.

File: DEVELOPER_GUIDE.md (hunk 1 above)
Comment (akashsha1) [reply DEVELOPER_GUIDE.md:294]: Ideally we want simd (avx2, avx512 - highest of whichever is present) to be enabled by default. The order of checks would be: 
if (AVX512 enabled and present) { use avx512 } 
else if (AVX2 enabled and present) { use avx2 }
else { use generic version }

by making it mutually exclusive, the intermediate step cannot be achieved.

Comment (naveentatikonda) [reply DEVELOPER_GUIDE.md:294]: I think Jack's point is to update it as something like this `./gradlew build -Davx2.enabled=false -Davx512.enabled=true` as we can't use both of them at the same time

File: src/main/java/org/opensearch/knn/index/KNNSettings.java (hunk 2 above)
Comment (akashsha1) [reply src/main/java/org/opensearch/knn/index/KNNSettings.java]: Looks like java boolean cannot be null. So null check won't be possible. 
Your second point on exceptions is valid, and this code shouldn't throw exceptions as a default value is set. I've removed the try/catch block.

Comment (naveentatikonda) [reply src/main/java/org/opensearch/knn/index/KNNSettings.java]: @akashsha1 as users will manually set this setting in opensearch.yml, there is a chance of getting null. To avoid it shall we change it to 
`return Booleans.parseBoolean(KNNSettings.state().getSettingValue(KNNSettings.KNN_FAISS_AVX512_DISABLED).toString(), false);`

Comment (naveentatikonda) [reply src/main/java/org/opensearch/knn/index/KNNSettings.java]: `Booleans.parseBoolean` will do the null validation and if it is null, it will return the default value

Comment (akashsha1) [reply src/main/java/org/opensearch/knn/index/KNNSettings.java]: spoke with Naveen on slack, and updated to 
return Booleans.parseBoolean(KNNSettings.state().getSettingValue(KNNSettings.KNN_FAISS_AVX512_DISABLED).toString(), KNN_DEFAULT_FAISS_AVX512_DISABLED_VALUE);

**Review by jmazanec15 — APPROVED** [review]
LGTM thanks

**Review by naveentatikonda — APPROVED** [review]
